src/core/Akka/Actor/ActorCell.DefaultMessages.cs

```diff
@@ -246,90 +246,66 @@ private void SendAllToDeadLetters(EarliestFirstSystemMessageList messages)
-                    }
-                }
-                catch (Exception cause)
+                switch (message)
                 {
-                    HandleInvokeFailure(cause);
+                    case SystemMessage sm when ShouldStash(sm, currentState):
+                        Stash(message);
+                        break;
+                    case ActorTaskSchedulerMessage atsm:
+                        HandleActorTaskSchedulerMessage(atsm);
+                        break;
+                    case Failed f:
+                        HandleFailed(f);
+                        break;
+                    case DeathWatchNotification n:
+                        WatchedActorTerminated(n.Actor, n.ExistenceConfirmed, n.AddressTerminated);
+                        break;
+                    case Create c:
+                        Create(c.Failure);
+                        break;
+                    case Watch w:
+                        AddWatcher(w.Watchee, w.Watcher);
+                        break;
+                    case Unwatch uw:
+                        RemWatcher(uw.Watchee, uw.Watcher);
+                        break;
+                    case Recreate r:
+                        FaultRecreate(r.Cause);
+                        break;
+                    case Suspend _:
+                        FaultSuspend();
+                        break;
+                    case Resume r:
+                        FaultResume(r.CausedByFailure);
+                        break;
+                    case Terminate _:
+                        Terminate();
+                        break;
+                    case Supervise s:
+                        Supervise(s.Child, s.Async);
+                        break;
+                    default:
+                        throw new NotSupportedException($"Unknown message {message.GetType().Name}");
                 }
-
-                nextState = CalculateState();
-                // As each state accepts a strict subset of another state, it is enough to unstash if we "walk up" the state
-                // chain
-                todo = nextState < currentState ? todo + UnstashAll() : todo;
-                if (IsTerminated)
-                {
-                    SendAllToDeadLetters(todo);
-                    return;
-                }
-                if (todo.IsEmpty) return; // keep running until the stash is empty
             }
+            catch (Exception cause)
+            {
+                HandleInvokeFailure(cause);
+            }
+
+            var newState = CalculateState();
+            // As each state accepts a strict subset of another state, it is enough to unstash if we "walk up" the state
+            // chain
+            var todo = newState < currentState ? rest + UnstashAll() : rest;
 
+            if (IsTerminated) SendAllToDeadLetters(todo);
+            else if (!todo.IsEmpty) SysMsgInvokeAll(todo, newState);
```

Comment:
IMHO, let's put the `while` loop back in place so we can save on recursion costs - shouldn't be a big deal either way since system messages are pretty rare but I'd feel more comfortable only changing one thing at a time here.

Reply:
done